File: backend/src/utils/pushNotification.js
import admin from "firebase-admin";
import path from "path";
import { fileURLToPath } from "url";
import { v4 as uuidv4 } from "uuid";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const serviceAccountPath = path.resolve(
  __dirname,
  "../config/push-notification-key.json"
);

if (!admin.apps.length) {
  admin.initializeApp({
    credential: admin.credential.cert(serviceAccountPath),
  });
}

//topup notification
export const sendTopupNotification = async (
  fcmToken,
  points,
  amount,
  firstName
) => {
  try {
    const uuid = uuidv4();
    const formattedAmount = new Intl.NumberFormat("id-ID", {
      style: "currency",
      currency: "IDR",
      minimumFractionDigits: 2, // Pastikan 2 digit desimal
      maximumFractionDigits: 2,
    }).format(amount / 100); // BAGI DENGAN 100

    const message = {
      notification: {
        title: `Hai ${firstName}, Top Up Berhasil! 🎉`,
        body: `Anda berhasil top up ${points} poin senilai ${formattedAmount}.`,
      },
      data: {
        title: `Hai ${firstName}, Top Up Berhasil! 🎉`,
        body: `Anda berhasil top up ${points} poin senilai ${formattedAmount}.`,
        type: "topup_success",
        points: points.toString(),
        amount: amount.toString(),
        uuid: uuid,
        click_action: "FLUTTER_NOTIFICATION_CLICK",
      },
      android: {
        priority: "high",
        notification: {
          channel_id: "topup_channel",
          sound: "default",
          tag: uuid,
        },
      },
      token: fcmToken,
    };

    return await admin.messaging().send(message);
  } catch (error) {
    console.error("Gagal mengirim notifikasi topup:", error);
    throw error;
  }
};

//pesanan cod notification
export const sendOrderCODNotification = async (
  fcmToken,
  orderId,
  totalAmount,
  firstName
) => {
  try {
    const uuid = uuidv4();
    const formattedAmount = new Intl.NumberFormat("id-ID", {
      style: "currency",
      currency: "IDR",
    }).format(totalAmount);

    const message = {
      notification: {
        title: `Hai ${firstName}, Pesanan COD Berhasil 🎉`,
        body: `Terima kasih telah berbelanja! Pesanan #${orderId} senilai ${formattedAmount} sudah kami terima.`,
      },
      data: {
        title: `Hai ${firstName}, Pesanan COD Berhasil 🎉`,
        body: `Terima kasih telah berbelanja! Pesanan #${orderId} senilai ${formattedAmount} sudah kami terima.`,
        type: "new_order",
        orderId: orderId.toString(),
        uuid: uuid,
        click_action: "FLUTTER_NOTIFICATION_CLICK",
      },
      android: {
        priority: "high",
        notification: {
          channel_id: "order_channel",
          sound: "default",
          tag: uuid,
        },
      },
      token: fcmToken,
    };

    return await admin.messaging().send(message);
  } catch (error) {
    console.error("Gagal mengirim notifikasi:", error);
    throw error;
  }
};

//pesanan poin notification
export const sendOrderPOinNotification = async (
  fcmToken,
  orderId,
  totalAmount,
  firstName
) => {
  try {
    const uuid = uuidv4();
    const message = {
      notification: {
        title: `Hai ${firstName}, Pesanan POIN Berhasil 🎉`,
        body: `Terima kasih telah berbelanja! Pesanan #${orderId} senilai ${totalAmount} Poin sudah kami terima.`,
      },
      data: {
        title: `Hai ${firstName}, Pesanan POIN Berhasil 🎉`,
        body: `Terima kasih telah berbelanja! Pesanan #${orderId} senilai ${totalAmount} Poin sudah kami terima.`,
        type: "new_order",
        orderId: orderId.toString(),
        uuid: uuid,
        click_action: "FLUTTER_NOTIFICATION_CLICK",
      },
      android: {
        priority: "high",
        notification: {
          channel_id: "order_channel",
          sound: "default",
          tag: uuid,
        },
      },
      token: fcmToken,
    };

    return await admin.messaging().send(message);
  } catch (error) {
    console.error("Gagal mengirim notifikasi:", error);
    throw error;
  }
};

export const sendStatusNotification = async (fcmToken, orderId, status) => {
  try {
    const uuid = uuidv4();
    const formattedStatus = status.charAt(0).toUpperCase() + status.slice(1);

    const message = {
      notification: {
        title: `Status Pesanan #${orderId} Diperbarui`,
        body: `Pesanan Anda sekarang dalam status: ${formattedStatus}`,
      },
      data: {
        title: `Status Pesanan #${orderId} Diperbarui`,
        body: `Pesanan Anda sekarang dalam status: ${formattedStatus}`,
        type: "status_update",
        orderId: orderId.toString(),
        uuid: uuid,
        click_action: "FLUTTER_NOTIFICATION_CLICK",
      },
      android: {
        priority: "high",
        notification: {
          channel_id: "status_channel",
          sound: "default",
          tag: uuid,
        },
      },
      token: fcmToken,
    };

    return await admin.messaging().send(message);
  } catch (error) {
    console.error("Gagal mengirim notifikasi status:", error);
    throw error;
  }
};

export const isFcmTokenValid = async (fcmToken) => {
  try {
    // Coba kirim pesan test (dryRun: true tidak benar-benar mengirim)
    await admin.messaging().send(
      {
        token: fcmToken,
        data: { validation: "true" },
      },
      true // dryRun mode
    );
    return true;
  } catch (error) {
    console.error(`Token validation failed: ${fcmToken}`, error);
    return false;
  }
};

File: backend-go/utils/push_notification.go
package utils

import (
	"context"
	"log"
	"strconv"
	"strings"
	"sync"

	firebase "firebase.google.com/go"
	"firebase.google.com/go/messaging"
	"github.com/google/uuid"
	"golang.org/x/text/language"
	"golang.org/x/text/message"
	"google.golang.org/api/option"
)

var (
	app     *firebase.App
	once    sync.Once
	errInit error
)

// Inisialisasi Firebase (hanya sekali)
func initFirebase() {
	credFilePath := "config/firebase-service-account.json"
	opt := option.WithCredentialsFile(credFilePath)
	once.Do(func() {
		ctx := context.Background()
		var err error
		app, err = firebase.NewApp(ctx, nil, opt)
		if err != nil {
			errInit = err
		}
	})
}

// Mendapatkan client messaging
func getMessagingClient() (*messaging.Client, error) {
	initFirebase()
	if errInit != nil {
		return nil, errInit
	}
	ctx := context.Background()
	return app.Messaging(ctx)
}

// Format mata uang IDR
func formatIDR(amount float64) string {
	p := message.NewPrinter(language.Indonesian)
	formattedNumber := p.Sprintf("%.0f", amount)

	return "Rp " + strings.ReplaceAll(formattedNumber, ",", ".")
}

// Notifikasi Topup
func SendTopupNotification(fcmToken string, points int, amount int, firstName string) {
	if fcmToken == "" {
		return
	}

	amountFloat := float64(amount) / 100.0
	formattedAmount := formatIDR(amountFloat)
	uuidVal := uuid.New().String()

	title := "Hai " + firstName + ", Top Up Berhasil! 🎉"
	body := "Anda berhasil top up " + strconv.Itoa(points) + " poin senilai " + formattedAmount + "."

	msg := &messaging.Message{
		Token: fcmToken,
		Notification: &messaging.Notification{
			Title: title,
			Body:  body,
		},
		Data: map[string]string{
			"title":        title,
			"body":         body,
			"type":         "topup_success",
			"points":       strconv.Itoa(points),
			"amount":       strconv.Itoa(amount),
			"uuid":         uuidVal,
			"click_action": "FLUTTER_NOTIFICATION_CLICK",
		},
		Android: &messaging.AndroidConfig{
			Priority: "high",
			Notification: &messaging.AndroidNotification{
				ChannelID: "topup_channel",
				Sound:     "default",
				Tag:       uuidVal,
			},
		},
	}

	sendMessage(msg)
}

// Notifikasi Order COD
func SendOrderCODNotification(fcmToken, orderId string, totalAmount int, firstName string) {
	if fcmToken == "" {
		return
	}

	formattedAmount := formatIDR(float64(totalAmount))
	uuidVal := uuid.New().String()

	title := "Hai " + firstName + ", Pesanan COD Berhasil 🎉"
	body := "Terima kasih telah berbelanja! Pesanan #" + orderId + " senilai " + formattedAmount + " sudah kami terima."

	msg := &messaging.Message{
		Token: fcmToken,
		Notification: &messaging.Notification{
			Title: title,
			Body:  body,
		},
		Data: map[string]string{
			"title":        title,
			"body":         body,
			"type":         "new_order",
			"orderId":      orderId,
			"uuid":         uuidVal,
			"click_action": "FLUTTER_NOTIFICATION_CLICK",
		},
		Android: &messaging.AndroidConfig{
			Priority: "high",
			Notification: &messaging.AndroidNotification{
				ChannelID: "order_channel",
				Sound:     "default",
				Tag:       uuidVal,
			},
		},
	}

	sendMessage(msg)
}

// Notifikasi Order Poin
func SendOrderPoinNotification(fcmToken, orderId string, totalAmount int, firstName string) {
	if fcmToken == "" {
		return
	}

	uuidVal := uuid.New().String()

	title := "Hai " + firstName + ", Pesanan POIN Berhasil 🎉"
	body := "Terima kasih telah berbelanja! Pesanan #" + orderId + " senilai " + strconv.Itoa(totalAmount) + " Poin sudah kami terima."

	msg := &messaging.Message{
		Token: fcmToken,
		Notification: &messaging.Notification{
			Title: title,
			Body:  body,
		},
		Data: map[string]string{
			"title":        title,
			"body":         body,
			"type":         "new_order",
			"orderId":      orderId,
			"uuid":         uuidVal,
			"click_action": "FLUTTER_NOTIFICATION_CLICK",
		},
		Android: &messaging.AndroidConfig{
			Priority: "high",
			Notification: &messaging.AndroidNotification{
				ChannelID: "order_channel",
				Sound:     "default",
				Tag:       uuidVal,
			},
		},
	}

	sendMessage(msg)
}

// Notifikasi Update Status
func SendStatusNotification(fcmToken, orderId, status string) {
	if fcmToken == "" {
		return
	}

	formattedStatus := status
	if len(status) > 0 {
		formattedStatus = strings.ToUpper(status[:1]) + strings.ToLower(status[1:])
	}

	uuidVal := uuid.New().String()

	title := "Status Pesanan #" + orderId + " Diperbarui"
	body := "Pesanan Anda sekarang dalam status: " + formattedStatus

	msg := &messaging.Message{
		Token: fcmToken,
		Notification: &messaging.Notification{
			Title: title,
			Body:  body,
		},
		Data: map[string]string{
			"title":        title,
			"body":         body,
			"type":         "status_update",
			"orderId":      orderId,
			"uuid":         uuidVal,
			"click_action": "FLUTTER_NOTIFICATION_CLICK",
		},
		Android: &messaging.AndroidConfig{
			Priority: "high",
			Notification: &messaging.AndroidNotification{
				ChannelID: "status_channel",
				Sound:     "default",
				Tag:       uuidVal,
			},
		},
	}

	sendMessage(msg)
}

// Validasi FCM Token
func IsFcmTokenValid(fcmToken string) bool {
	if fcmToken == "" {
		return false
	}

	client, err := getMessagingClient()
	if err != nil {
		log.Printf("Failed to get FCM client: %v", err)
		return false
	}

	msg := &messaging.Message{
		Token: fcmToken,
		Data:  map[string]string{"validation": "true"},
	}

	_, err = client.SendDryRun(context.Background(), msg)
	return err == nil
}

// Fungsi internal pengirim pesan
func sendMessage(msg *messaging.Message) {
	client, err := getMessagingClient()
	if err != nil {
		log.Printf("Failed to get FCM client: %v", err)
		return
	}

	_, err = client.Send(context.Background(), msg)
	if err != nil {
		log.Printf("Failed to send FCM message: %v", err)
	}
}
